patroni/postgresql.py

```diff
@@ -622,12 +658,84 @@ def reload(self):
             self.call_nowait(ACTION_ON_RELOAD)
         return ret
 
-    def restart(self):
-        self.set_state('restarting')
-        ret = self.stop(block_callbacks=True) and self.start(block_callbacks=True)
-        if ret:
-            self.call_nowait(ACTION_ON_RESTART)
+    def check_for_startup(self):
+        """Checks PostgreSQL status and returns if PostgreSQL is in the middle of startup."""
+        if self.is_starting():
+            return not self.check_startup_state_changed()
         else:
+            ready = self.pg_isready()
+            if ready == STATE_REJECT:
+                # Maybe add a ACTION_ON_FAIL here?
+                self.set_state('starting')
+                return True
+            else:
+                # Other cases will be handled by is_healthy check
+                return False
+
+    def check_startup_state_changed(self):
+        """Checks if PostgreSQL has completed starting up or failed or still starting.
+
+        Should only be called when state == 'starting'
+
+        :returns: True iff state was changed from 'starting'
+        """
+        ready = self.pg_isready()
+
+        if ready == STATE_REJECT:
+            return False
+        elif ready == STATE_NO_RESPONSE:
```

Comment:
We can't reliably distinguish between the two cases when using pg\_ctl as it doesn't return the pid that was started. That said, there is very little processing before opening ports. Some system checks, config file loading, pid file creation and `shared_preload_libraries`.  I added some code to do the waiting for port open from within Patroni and increased the timeout. If ports can't be opened within the given timeout I guess it's not too far from the truth to say that start failed.

Next iterations of run_cycle will then either notice that PostgreSQL moved into the "starting" state, see that it successfully started, or retry starting it, which will race with the initial invocation to acquire data dir lock file. So if this is a temporary issue (swap storm, VM move/pause), it will resolve itself eventually. If it isn't there isn't much we can do anyway.

Alternative would be to start postmaster directly. Then we know the pid that we started and can check that it's running even when it hasn't gotten around to creating the pid file yet. I don't think it helps the correctly configured use case too much. But it would make detection of wrong permissions on data dir faster as we notice immediately that the process we started went away without creating a pid file.